### PR TITLE
[Autoscaling] Add autoscaling e2e test

### LINKED ISSUE #4276: [Autoscaling] E2E testing
We should validate that the autoscaling mechanism is working as expected with some e2e tests.

If we don't want to rely on some external tooling (I think we should not) an option would be [to use the fixed decider](https://www.elastic.co/guide/en/elasticsearch/reference/current/autoscaling-fixed-decider.html).

### PR DESCRIPTION
**Pull request by barkbay**
Fixes #4276 

This new e2e test:

* Initially deploys a cluster with 2 data nodes and an empty ML tier.
* Scale up vertically and horizontally the data nodes to 3 instances.
* Scale up the ML tier to 1 node.
* Scale down the ML tier to 0 node.

Test duration is ~6 minutes from my laptop, tested on `GKE` and `OCP4`.

I had to introduce a new field named `expectedElasticsearch` in the Elasticsearch `Builder`. It might sound a bit odd to have an "expected" state in a `Builder`, but as mentioned in #2098 builders are not only builders: they are also used during the test execution phase. It becomes a problem when the Elasticsearch resource described in the Builder is not the one actually deployed, as it is the case when the resource is "externally" modified by the autoscaling controller.

### CODE REVIEW

**Review by pebrc — COMMENTED** [review]
LGTM, nice that this led to a bit of cleanup around the ES checks code! I think the `expectedElasticsearch` thingy is not too bad. 👍

File: test/e2e/es/autoscaling_test.go [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+			// Autoscaling controller should eventually update the data node count to its min. value.
+			newExpectedNodeSet("data-ingest", []string{"data", "ingest"}, 2, corev1.ResourceList{corev1.ResourceMemory: nodespec.DefaultMemoryLimits}),
+			// ML node count should still be 0.
+			newExpectedNodeSet("ml", []string{"ml"}, 0, corev1.ResourceList{}),
+		)
+
+	// scaleUpStorage uses the fixed decider to trigger a scale up of the data tier up to its max memory limit and 3 nodes.
+	scaleUpStorage := esBuilder.DeepCopy().WithAnnotation(
+		esv1.ElasticsearchAutoscalingSpecAnnotationName,
+		autoscalingSpecBuilder.withFixedDecider("data-ingest", map[string]string{"storage": "20gb", "nodes": "3"}).toJSON(),
```
[the remark this answers is not in the document]
Comment (barkbay) [reply test/e2e/es/autoscaling_test.go:81]: You're right, [CheckESNodesTopology](https://github.com/elastic/cloud-on-k8s/blob/2a781357091e87d3a026d08bb3fa8ead43d67b9b/test/e2e/test/elasticsearch/checks_es.go#L88) only validates the memory settings.

I think we should improve the resources validation but I'm wondering if it is a good idea to do that in this PR as it's impacting all the other tests ? I think I would rather fix it in a dedicated PR. Let me know if you think it's fine to do that.

[the remark this answers is not in the document]
Comment (pebrc) [reply test/e2e/es/autoscaling_test.go:81]: Yeah that's fine. Let's maybe add a comment to be explicit about what is being verified.

**Review by pebrc — APPROVED** [review]
LGTM!

**Comment by barkbay** [comment]
Jenkins test this please

**Pull request merged**